packages/diagrams/frontend/sirius-components-diagrams-reactflow/src/renderer/node/ConnectionSourceHandles.tsx

```diff
@@ -0,0 +1,260 @@
+ * Copyright (c) 2023 Obeo and others.
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Obeo - initial API and implementation
+ *******************************************************************************/
+import { gql, useQuery } from '@apollo/client';
+import { useMultiToast } from '@eclipse-sirius/sirius-components-core';
+import { Theme, useTheme } from '@material-ui/core/styles';
+import React, { useContext, useEffect, useState } from 'react';
+import { Handle, Position, useUpdateNodeInternals } from 'reactflow';
+import { DiagramContext } from '../../contexts/DiagramContext';
+import { DiagramContextValue } from '../../contexts/DiagramContext.types';
+import { ConnectorContext } from '../connector/ConnectorContext';
+import { ConnectorContextValue } from '../connector/ConnectorContext.types';
+import { useConnector } from '../connector/useConnector';
+import {
+  GQLDiagramDescription,
+  GQLGetToolSectionsData,
+  GQLGetToolSectionsVariables,
+  GQLNodeDescription,
+  GQLRepresentationDescription,
+  GQLSingleClickOnTwoDiagramElementsTool,
+  GQLTool,
+} from '../connector/useConnector.types';
+import { ConnectionSourceHandlesProps, ConnectionSourceHandlesStates } from './ConnectionSourceHandles.types';
+
+const getToolSectionsQuery = gql`
+  query getToolSections($editingContextId: ID!, $diagramId: ID!, $diagramElementId: ID!) {
+    viewer {
+      editingContext(editingContextId: $editingContextId) {
+        representation(representationId: $diagramId) {
+          description {
+            ... on DiagramDescription {
+              palette(diagramElementId: $diagramElementId) {
+                tools {
+                  __typename
+                  ... on SingleClickOnTwoDiagramElementsTool {
+                    candidates {
+                      sources {
+                        id
+                      }
+                      targets {
+                        id
+                      }
+                    }
+                  }
+                }
+                toolSections {
+                  tools {
+                    __typename
+                    ... on SingleClickOnTwoDiagramElementsTool {
+                      candidates {
+                        sources {
+                          id
+                        }
+                        targets {
+                          id
+                        }
+                      }
+                    }
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+`;
+
+const customConnectionHandlesStyle = (position: Position, theme: Theme): React.CSSProperties => {
+  switch (position) {
```

Comment:
Difficult to read what's the same (most properties) and what is different between the 4 cases.